alea/test/stream_serializer.cpp

```diff
@@ -0,0 +1,141 @@
+#include <alps/alea/plugin/stream_serializer.hpp>
+
+#include "gtest/gtest.h"
+#include "dataset.hpp"
+
+#include <vector>
+#include <queue>
+#include <cstdint>
+
+// Mock class that emulates Boost/HPX serialization archive
+class MockArchive {
+
+public:
+
+    MockArchive() {}
+
+    // Store simple types
+    MockArchive & operator<<(double x) { store_fundamental(x); return *this; }
+    MockArchive & operator<<(long x) { store_fundamental(x); return *this; }
+    MockArchive & operator<<(unsigned long x) { store_fundamental(x); return *this; }
+    MockArchive & operator<<(std::complex<double> x)
+    {
+        *this << x.real() << x.imag();
+        return *this;
+    }
+
+    // Store complex_op<T>
+    template<typename T>
+    MockArchive & operator<<(const alps::alea::complex_op<T> &x)
+    {
+        *this << x.rere() << x.reim() << x.imre() << x.imim();
+        return *this;
+    }
+
+    // Store ALEA results
+    template <typename T>
+    typename std::enable_if<alps::alea::is_alea_result<T>::value, MockArchive &>::type
+    operator<<(const T &result)
+    {
+        save(*this, result, 0);
+        return *this;
+    }
+
+    // Extract simple types
+    MockArchive & operator>>(double &x) { extract_fundamental(x); return *this; }
+    MockArchive & operator>>(long &x) { extract_fundamental(x); return *this; }
+    MockArchive & operator>>(unsigned long &x) { extract_fundamental(x); return *this; }
+    MockArchive & operator>>(std::complex<double> &x) {
+      double r, i;
+      extract_fundamental(r);
+      extract_fundamental(i);
+      x = std::complex<double>(r, i);
+    }
+
+    // Extract complex_op<T>
+    template<typename T>
+    MockArchive & operator>>(alps::alea::complex_op<T> &x)
+    {
+        *this >> x.rere() >> x.reim() >> x.imre() >> x.imim();
+        return *this;
+    }
+
+    // Extract ALEA results
+    template <typename T>
+    typename std::enable_if<alps::alea::is_alea_result<T>::value, MockArchive &>::type
+    operator>>(T &result)
+    {
+        load(*this, result, 0);
+        return *this;
+    }
+
+private:
+
+    template<typename T>
+    void store_fundamental(T x) {
+        std::int8_t* p = reinterpret_cast<std::int8_t*>(&x);
+        for(int n = 0; n < sizeof(T); ++n)
+            buf.push(*(p + n));
+    }
+
+    template<typename T>
+    void extract_fundamental(T &x) {
+        std::int8_t* p = reinterpret_cast<std::int8_t*>(&x);
```

Comment:
TL;DR;  At the risk of sounding pedantic, i suggest using `unsigned char` rather than `std::int8_t`

Rationale:
According to [type aliasing rules](http://en.cppreference.com/w/cpp/language/reinterpret_cast#Type_aliasing), accessing `x` via `p` is UB, unless `std::int8_t` happens to be a typedef for `unsigned char`, and it cannot be the case, because according to  http://en.cppreference.com/w/cpp/types/integer, `std::int8_t` is a signed type.

(As a side-note, I believe `std::uint8_t` is most probably a typedef for `unsigned char` on architectures we care about, although those "fixed width integer types" are technically optional.)

But we should use `unsigned char` regardless, because we really do not care if our `char` is 8 bits --- `sizeof` is defined in terms of `char`s, not 8-bit bytes.